clients/spark/core/src/main/scala/io/treeverse/clients/StorageUtils.scala

```diff
@@ -80,5 +87,71 @@ object StorageUtils {
       if (keys.isEmpty) return Seq.empty
       keys.map(x => snPrefix.concat(x))
     }
+
+    def createAndValidateS3Client(
+        configuration: ClientConfiguration,
+        credentialsProvider: Option[AWSCredentialsProvider],
+        awsS3ClientBuilder: AmazonS3ClientBuilder,
+        region: String,
+        bucket: String
+    ): AmazonS3 = {
+
+      require(awsS3ClientBuilder != null)
+      require(bucket.nonEmpty)
+
+      var client =
+        initializeS3Client(configuration, credentialsProvider, awsS3ClientBuilder, region)
+
+      if (!validateClientAndBucketRegionsMatch(client, bucket)) {
+        val bucketRegion = getAWSS3Region(client, bucket)
+        client = initializeS3Client(configuration,
+                                    credentialsProvider,
+                                    AmazonS3ClientBuilder.standard(),
+                                    bucketRegion
+                                   )
+      }
+      client
+    }
+
+    private def initializeS3Client(
+        configuration: ClientConfiguration,
+        credentialsProvider: Option[AWSCredentialsProvider],
+        awsS3ClientBuilder: AmazonS3ClientBuilder,
+        region: String
+    ): AmazonS3 = {
+      val builder = awsS3ClientBuilder
+        .withClientConfiguration(configuration)
+        .withRegion(region)
+      val builderWithCredentials = credentialsProvider match {
+        case Some(cp) => builder.withCredentials(cp)
+        case None     => builder
+      }
+      builderWithCredentials.build
```

Comment:
I think there is some Scala style that recommends using bracket-less calls _only_ for code with no side-effects.  `Builder.build` returns a `new` object, so (I think) this would be clearer as
```suggestion
      builderWithCredentials.build()
```

Reply:
that was actually an IntelliJ suggestion 🤷‍♂️